Sources/Commands/SwiftTestTool.swift

```diff
@@ -320,6 +303,182 @@ public class SwiftTestTool: SwiftTool<TestMode, TestToolOptions> {
+      #else
+        args += [path.asString]
+        if let xctestArg = xctestArg {
+            args += [xctestArg]
+        }
+      #endif
+        return args
+    }
+
+    /// Current inherited enviornment variables.
+    private var environment: [String: String] {
+        return ProcessInfo.processInfo.environment
+    }
+
+    /// Executes the tests without printing anything on standard streams.
+    ///
+    /// - Returns: A tuple with first bool member indicating if test execution returned code 0
+    ///            and second argument containing the output of the execution.
+    func test() -> (Bool, String) {
+        var output = ""
+        var success = true
+        do {
+            try popen(args(), redirectStandardError: true, environment: environment) { line in
+                output += line
+            }
+        } catch {
+            success = false
+        }
+        return (success, output)
+    }
+
+    /// Executes and returns execution status. Prints test output on standard streams.
+    func test() -> Bool {
+        let result: Void? = try? system(args(), environment: environment)
+        return result != nil
+    }
+}
+
+/// A class to run tests in parallel.
+final class ParallelTestRunner {
+    /// Path to XCTest binary.
+    private let testPath: AbsolutePath
+
+    /// The queue containing list of tests to run (producer).
+    private let pendingTests = SynchronizedQueue<String?>()
+
+    /// The queue containing output of failed tests.
+    private let failedTests = SynchronizedQueue<String?>()
+
+    /// The queue containing tests which are finished running.
+    private let finishedTests = SynchronizedQueue<String?>()
+
+    /// Number of parallel workers to spawn.
+    private var numJobs: Int {
+        return ProcessInfo.processInfo.activeProcessorCount
+    }
+
+    /// Instance of progress bar. Animating progress bar if stream is a terminal otherwise
+    /// a simple progress bar.
+    private let progressBar: ProgressBarProtocol
+
+    /// Number of tests that will be executed.
+    private let numTests: Int
+
+    /// Number of the current tests that has been executed.
+    private var numCurrentTest = 0
+
+    /// True if all tests executed successfully.
+    private(set) var success: Bool = true
+
+    init(testPath: AbsolutePath) throws {
+        self.testPath = testPath
+        progressBar = createProgressBar(forStream: stdoutStream, header: "Tests")
+        // Find all the test suites present in the test binary.
+        let testSuites = try SwiftTestTool.getTestSuites(path: testPath)
+        // FIXME: Add a count property in SynchronizedQueue.
+        var numTests = 0
+        // Enqueue all the tests.
+        for testSuite in testSuites {
+            for testCase in testSuite.tests {
+                for test in testCase.tests {
+                    numTests += 1
+                    pendingTests.enqueue(testCase.name + "/" + test)
+                }
+            }
+        }
```

Comment:
It feels weird to initialize the pendingTests queue in the initializer but consume it inside run(). Any reason not to just to it in run?